### ISSUE TITLE
Allow users to reset MFA

### ISSUE THREAD

**Issue by austinbyers** [panther-labs/panther#317]
### Describe the ideal solution
Users can update their own MFA (e.g. when they get a new phone). Admins should be able to reset MFA as well as passwords

### Describe your use cases
If a user gets a new phone, they might want to setup MFA on the new phone. Also, if someone loses their MFA device, admins should have a way to reset them without having to delete and re-add their user

### How frequently would you use such feature

Occasionally

**Comment by 3nvi** [comment]
It's not as easy. If I remember correctly, Cognito doesn't give you this ability, so you need another way to achieve that

**Comment by 3nvi** [comment]
@austinbyers 

From [here](https://stackoverflow.com/questions/50497388/aws-cognito-reset-user-mfa):

>At this point, since AWS does not support resetting the MFA (if your user pool requires MFA - disabling MFA using AdminSetUserMFAPreference will return 200 OK but it will do nothing), the only way to do this is to create a new user pool with optional MFA (you have to create a new one since changing from required to optional is prohibited once the user pool is created). Then with the new user pool, you have to enforce the MFA (if that's something that you want) manually within your code.

> For that, once the user logs in successfully and the return object has the tokens in it, you have to call AssociateSoftwareToken instead of returning the tokens and start the MFA registration process. The tokens (like IdToken) will only be returned to the user when they managed to complete the call to AdminRespondToAuthChallenge successfully.

> Bottom line, with the optional MFA, the AdminSetUserMFAPreference will work. And this is the only way to reset MFA in a user pool on AWS (at this point).

**Comment by austinbyers** [comment]
Good to know! Seems like there is an option, but a technically challenging one. I'll close as "won't fix

Enterprise users can always integrate SSO systems for more advanced MFA capability if they want them

**Comment by 3nvi** [comment]
@austinbyers Seems like currently, the only way we can achieve this is if the MFA is configured as `optional`. Then we can disable/re-enable MFA for each user and it will "mimic" the reset behavior. If we have to have MFA configured as `required` then this ticket can safely be closed.

I remember that at one point we did indeed have MFA set as `optional` in the Cognito settings, but we changed it to `required` for some reason. Do you have any more context?

**Comment by austinbyers** [comment]
> I remember that at one point we did indeed have MFA set as optional in the Cognito settings, but we changed it to required for some reason. Do you have any more context?

Ah, good question! No, I don't have context on why we did that. I didn't realize we had this option available to us previously, I'll reopen then until we do a bit more research

**Comment by 3nvi** [comment]
I spent some time investigating that. Indeed if we have MFA set to `Required` for the entire user pool, then Cognito doesn't give you any option to reset the MFA. I tried disabling it and re-enabling it for a user as well as the entire user pool and it didn't reset anything. The only way to do it is to follow the instructions in my comment above.

We don't want optional MFA (which would mean enabling manually through code for each registered user), since this may be restricting for certain compliance checks, while it would involve a lot of work and  "workarounds" to make it work. I'm closing this ticket Cognito clearly doesn't support this natively.
